flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTest.java

```diff
@@ -618,9 +837,75 @@ protected long getSizeInBytes(Integer requestEntry) {
         }
     }
 
+    /** A builder for {@link AsyncSinkWriterImpl}. */
+    public class AsyncSinkWriterImplBuilder {
+
+        private Boolean simulateFailures;
+        private int delay = 0;
+        private Sink.InitContext context;
+        private Integer maxBatchSize;
+        private Integer maxInFlightRequests;
+        private Integer maxBufferedRequests;
+        private long flushOnBufferSizeInBytes = 10000000;
+        private long maxTimeInBufferMS = 1000;
+
+        private AsyncSinkWriterImplBuilder context(Sink.InitContext context) {
+            this.context = context;
+            return this;
+        }
+
+        private AsyncSinkWriterImplBuilder maxBatchSize(int maxBatchSize) {
+            this.maxBatchSize = maxBatchSize;
+            return this;
+        }
+
+        private AsyncSinkWriterImplBuilder maxInFlightRequests(int maxInFlightRequests) {
+            this.maxInFlightRequests = maxInFlightRequests;
+            return this;
+        }
+
+        private AsyncSinkWriterImplBuilder maxBufferedRequests(int maxBufferedRequests) {
+            this.maxBufferedRequests = maxBufferedRequests;
+            return this;
+        }
+
+        private AsyncSinkWriterImplBuilder flushOnBufferSizeInBytes(long flushOnBufferSizeInBytes) {
+            this.flushOnBufferSizeInBytes = flushOnBufferSizeInBytes;
+            return this;
+        }
+
+        private AsyncSinkWriterImplBuilder maxTimeInBufferMS(long maxTimeInBufferMS) {
+            this.maxTimeInBufferMS = maxTimeInBufferMS;
+            return this;
+        }
+
+        private AsyncSinkWriterImplBuilder delay(int delay) {
+            this.delay = delay;
+            return this;
+        }
+
+        private AsyncSinkWriterImplBuilder simulateFailures(boolean simulateFailures) {
+            this.simulateFailures = simulateFailures;
+            return this;
+        }
+
+        private AsyncSinkWriterImpl build() {
+            return new AsyncSinkWriterImpl(
+                    context,
+                    maxBatchSize,
+                    maxInFlightRequests,
+                    maxBufferedRequests,
+                    flushOnBufferSizeInBytes,
+                    maxTimeInBufferMS,
+                    simulateFailures,
+                    delay);
+        }
+    }
+
     private static class SinkInitContext implements Sink.InitContext {
 
         private static final TestProcessingTimeService processingTimeService;
+        private final SinkWriterMetricGroupMock metricGroup = new SinkWriterMetricGroupMock();
```

Comment:
We have a `MetricListener` in Flink to ease testing of emitted metrics you take a look at https://github.com/apache/flink/blob/dabc1a894fd5bca1882a071b5d4a015f2d341a32/flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java#L153 how to use it.

I think by using it you can get rid of all the metric related mock classes.

Reply:
Thanks, it's allowed me to remove 2 classes.